internal/backend/runtime/omni/controllers/omni/cluster_machine_config.go

```diff
@@ -272,22 +272,33 @@ type clusterMachineConfigControllerHelper struct {
 	imageFactoryHost string
 }
 
-const talosVersionLatest = "latest"
-
 func (helper clusterMachineConfigControllerHelper) generateConfig(clusterMachine *omni.ClusterMachine, clusterMachineConfigPatches *omni.ClusterMachineConfigPatches, secrets *omni.ClusterSecrets,
 	loadbalancer *omni.LoadBalancerConfig, cluster *omni.Cluster, clusterConfigVersion *omni.ClusterConfigVersion, configGenOptions *omni.MachineConfigGenOptions, extraGenOptions []generate.Option,
 	connectionParams *siderolink.ConnectionParams, link *siderolink.Link, eventSinkPort int,
 ) ([]byte, error) {
 	clusterName := cluster.Metadata().ID()
 
-	talosVersion := cluster.TypedSpec().Value.TalosVersion
-	kubernetesVersion := clusterMachine.TypedSpec().Value.KubernetesVersion
-
-	if talosVersion == "" {
+	// this is the version of Talos at the moment the cluster got created
+	//
+	// [NOTE]: this should be kept a constant for the lifetime of the cluster,
+	// as it dictates the Talos machinery config generation defaults.
+	// If this value is changed, it will cause the machine configuration to be regenerated
+	// with new version contract (defaults), and might cause unexpected issues.
+	//
+	// The desired version of Talos for this machine (not for config generation), but for the
+	// e.g. install image is stored in MachineConfigGenOptions.
+	initialTalosVersion := clusterConfigVersion.TypedSpec().Value.Version
```

Comment:
this is the line of the actual fix, everything else is to simplify/make the code more clear so we don't make this mistake in the future